add-chain/cmd/compress-genesis.go

```diff
@@ -0,0 +1,255 @@
+			ExtraData:     genesis.ExtraData,
+			GasLimit:      genesis.GasLimit,
+			Difficulty:    (*hexutil.Big)(genesis.Difficulty),
+			Mixhash:       genesis.Mixhash,
+			Coinbase:      genesis.Coinbase,
+			Number:        genesis.Number,
+			GasUsed:       genesis.GasUsed,
+			ParentHash:    genesis.ParentHash,
+			BaseFee:       (*hexutil.Big)(genesis.BaseFee),
+			ExcessBlobGas: genesis.ExcessBlobGas, // EIP-4844
+			BlobGasUsed:   genesis.BlobGasUsed,   // EIP-4844
+			Alloc:         make(jsonutil.LazySortedJsonMap[common.Address, GenesisAccount]),
+		}
+
+		// write genesis, but only reference code by code-hash, and don't encode the L2 predeploys to save space.
+		for addr, account := range genesis.Alloc {
+			var codeHash common.Hash
+			if len(account.Code) > 0 {
+				codeHash = crypto.Keccak256Hash(account.Code)
+			}
+			outAcc := GenesisAccount{
+				CodeHash: codeHash,
+				Nonce:    account.Nonce,
+			}
+			if account.Balance != nil && account.Balance.Cmp(common.Big0) != 0 {
+				outAcc.Balance = (*hexutil.Big)(account.Balance)
+			}
+			if len(account.Storage) > 0 {
+				outAcc.Storage = make(jsonutil.LazySortedJsonMap[common.Hash, common.Hash])
+				for k, v := range account.Storage {
+					outAcc.Storage[k] = v
+				}
+			}
+			out.Alloc[addr] = outAcc
+		}
+
+		// write genesis alloc
+		if err := writeGzipJSON(zipOutputDir, out); err != nil {
+			return fmt.Errorf("failed to write output: %w", err)
+		}
+		return nil
+	},
+}
+
+func writeBytecode(bytecodesDir string, code []byte, addr common.Address) error {
+	codeHash := crypto.Keccak256Hash(code)
+	name := filepath.Join(bytecodesDir, fmt.Sprintf("%s.bin.gz", codeHash))
+	_, err := os.Stat(name)
+	if err == nil {
+		// file already exists
+		return nil
+	}
+	if !os.IsNotExist(err) {
+		return fmt.Errorf("failed to check for pre-existing bytecode %s for address %s: %w", codeHash, addr, err)
+	}
+	var buf bytes.Buffer
+	w, err := gzip.NewWriterLevel(&buf, 9)
+	if err != nil {
+		return fmt.Errorf("failed to construct gzip writer for bytecode %s: %w", codeHash, err)
+	}
+	if _, err := w.Write(code); err != nil {
+		return fmt.Errorf("failed to write bytecode %s to gzip writer: %w", codeHash, err)
+	}
+	if err := w.Close(); err != nil {
+		return fmt.Errorf("failed to close gzip writer: %w", err)
+	}
+	// new bytecode
+	if err := os.WriteFile(name, buf.Bytes(), 0o755); err != nil {
+		return fmt.Errorf("failed to write bytecode %s of account %s: %w", codeHash, addr, err)
+	}
+	fmt.Printf("created new bytecodes file: %s\n", filepath.Base(name))
+	return nil
+}
+
+type GenesisAccount struct {
+	CodeHash common.Hash                                          `json:"codeHash,omitempty"`
+	Storage  jsonutil.LazySortedJsonMap[common.Hash, common.Hash] `json:"storage,omitempty"`
+	Balance  *hexutil.Big                                         `json:"balance,omitempty"`
+	Nonce    uint64                                               `json:"nonce,omitempty"`
+}
+
+type Genesis struct {
```

Comment:
Can we get rid of this struct and use the existing `core.Genesis` one instead? I'm wondering if this was necessary to avoid a circular dependency when the code lived inside the monorepo but maybe is no longer needed

Reply:
I don't think there was a circular dependency. 

I think this is a different type on purpose because it represents the _minified_ version of the genesis, where code is replaced with codehash or storage is replaced with storage hash. https://github.com/ethereum-optimism/superchain-registry/blob/852e9cfe836ecff7bbe3a5d771b67acb20558391/README.md#L79-L83

Reply:
Perhaps you meant the type in the `superchain` module?

https://github.com/ethereum-optimism/superchain-registry/blob/52c557aa401e6aac03f528adb5090e4b61ccdc00/superchain/superchain.go#L761

We might be able to deduplicate against this one, although there might be something about ordering fields that we lose by doing that.

Reply:
Yea the superchain `Genesis` struct seems like the one we could deduplicate against. All of the fields seem identical except:
* superchain package reimplements its own types because it can't import go-ethereum (to avoid circular dependency)
  * `common.Hash` --> `Hash`
  * `common.Address` --> `Address`
  * `hexutil.Big` --> `HexBig`
  * `Alloc` is a `LazySortedJsonMap` vs a vanilla map. Not sure what the reasoning is for sorting the json map. Not sure its worth blocking this PR on it but maybe worth tracking down and optimizing by remove one of the redundant structs later?

Reply:
Perhaps the sorting helps with comparing the JSON files in a standard way? Not sure. 

Agree we shouldn't block this PR on it and we can return to it later.